libs/server/Storage/Session/MainStore/MainStoreOps.cs

```diff
@@ -1190,5 +1193,299 @@ public GarnetStatus MemoryUsageForKey<TContext, TObjectContext>(ArgSlice key, ou
+                    return GarnetStatus.OK;
+                }
+
+                ArgSlice val1, val2;
+                var status1 = GET(key1, out val1, ref context);
+                var status2 = GET(key2, out val2, ref context);
+
+                if (lenOnly)
+                {
+                    if (status1 != GarnetStatus.OK || status2 != GarnetStatus.OK)
+                    {
+                        while (!RespWriteUtils.WriteInteger(0, ref curr, end))
+                            ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                        return GarnetStatus.OK;
+                    }
+
+                    var len = ComputeLCSLength(val1.ReadOnlySpan, val2.ReadOnlySpan, minMatchLen);
+                    while (!RespWriteUtils.WriteInteger(len, ref curr, end))
+                        ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                }
+                else if (withIndices)
+                {
+                    List<LCSMatch> matches;
+                    int len;
+                    if (status1 != GarnetStatus.OK || status2 != GarnetStatus.OK)
+                    {
+                        matches = new List<LCSMatch>();
+                        len = 0;
+                    }
+                    else
+                    {
+                        matches = ComputeLCSWithIndices(val1.ReadOnlySpan, val2.ReadOnlySpan, minMatchLen, out len);
+                    }
+
+                    WriteLCSMatches(matches, withMatchLen, len, ref curr, end, ref output, ref isMemory, ref ptr, ref ptrHandle);
+                }
+                else
+                {
+                    if (status1 != GarnetStatus.OK || status2 != GarnetStatus.OK)
+                    {
+                        while (!RespWriteUtils.WriteDirect(CmdStrings.RESP_EMPTY, ref curr, end))
+                            ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                        return GarnetStatus.OK;
+                    }
+
+                    var lcs = ComputeLCS(val1.ReadOnlySpan, val2.ReadOnlySpan, minMatchLen);
+                    while (!RespWriteUtils.WriteBulkString(lcs, ref curr, end))
+                        ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                }
+            }
+            finally
+            {
+                if (isMemory)
+                    ptrHandle.Dispose();
+                output.Length = (int)(curr - ptr);
+            }
+
+            return GarnetStatus.OK;
+        }
+
+        private static int ComputeLCSLength(ReadOnlySpan<byte> str1, ReadOnlySpan<byte> str2, int minMatchLen)
+        {
+            var m = str1.Length;
+            var n = str2.Length;
+            var dp = new int[m + 1, n + 1];
```

Comment:
As you're only using the current and previous row, it's sufficient to hold 2 1D arrays.
In general, it is not a good idea to use multidimensional arrays, see: https://stackoverflow.com/questions/597720/differences-between-a-multidimensional-array-and-an-array-of-arrays

Reply:
I tried with 2D array first, but when using 2 1D arrays I am getting wrong results compared to Redis because it can't maintain the direction when backtracking. This implementation is inspired from https://www.geeksforgeeks.org/longest-common-subsequence-dp-4/. What I can do is that if we just need length (lenOnly option) then I can use 2D array, for others we need multidimensional arrays. I tried a lot with 2 1D arrays, I am pretty sure we can't use it when we need actual LCS back or its indices.

Reply:
I haven't looked deeply into this, but you try and can use Redis's implementation as reference. Please look here https://github.com/redis/redis/blob/unstable/src/t_string.c at `lcsCommand`

Reply:
Verified the valkey code that also uses multi-dimensional array with regressive calls, I think that might be less performant that our implementation.

Reply:
Link to the code?

Reply:
Actually that's not what they're doing here, they're compacting the matrix and storing only the values necessary for the computation.

Reply:
Let's perf this as Badrish suggested, we can compare with valkey for funsies

Reply:
I tested by adding BDN benchmark, `[m,n]`, `[m][n]`, `[m*n]` doesn't make any noticeable difference

Reply:
Any update on this?